core/src/main/scala/org/apache/spark/Dependency.scala

```diff
@@ -85,14 +84,28 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
   private[spark] val combinerClassName: Option[String] =
     Option(reflect.classTag[C]).map(_.runtimeClass.getName)
 
-  val shuffleId: Int = _rdd.context.newShuffleId()
-
   val shuffleHandle: ShuffleHandle = _rdd.context.env.shuffleManager.registerShuffle(
     shuffleId, _rdd.partitions.size, this)
+}
+
+
+private[spark] abstract class BaseShuffleDependency[T: ClassTag](
+    @transient private val _rdd: RDD[T],
+    val numPartitions: Int) extends Dependency[T] {
+
+  override def rdd: RDD[T] = _rdd
+
+  val shuffleId: Int = _rdd.context.newShuffleId()
 
   _rdd.sparkContext.cleaner.foreach(_.registerShuffleForCleanup(this))
 }
 
+private[spark] class BinaryShuffleDependency[T: ClassTag](
```

Comment:
This is the new dependency for use by the binary shuffle. It's only used to capture the structure of the wide dependency and to register a handler for the write-side of the shuffle.